cincinnati/src/plugins/internal/cincinnati_graph_fetch.rs

```diff
@@ -0,0 +1,281 @@
+//! This plugin implements fetching a Cincinnati graph via HTTP from a `/v1/graph`-compliant endpoint.
+
+extern crate futures;
+extern crate quay;
+extern crate tokio;
+
+use crate::plugins::{InternalIO, InternalPlugin, InternalPluginWrapper, PluginSettings};
+use failure::Fallible;
+use plugins::AsyncIO;
+use plugins::PluginReference;
+use prometheus::{Counter, Registry};
+use std::fmt::{Debug, Formatter};
+
+/// Default URL to upstream graph provider.
+pub static DEFAULT_UPSTREAM_URL: &str = "http://localhost:8080/v1/graph";
+
+/// Plugin settings.
+#[derive(Clone, Debug, Deserialize, SmartDefault)]
+#[serde(default)]
+struct CincinnatiGraphFetchSettings {
+    #[default(DEFAULT_UPSTREAM_URL.to_string())]
+    upstream: String,
+}
+
+/// Graph fetcher for Cincinnati `/v1/graph` endpoints.
+pub struct CincinnatiGraphFetchPlugin {
+    pub upstream: String,
+    pub prometheus_registry: Option<Registry>,
+    pub http_upstream_reqs: Option<Counter>,
+    pub http_upstream_unreachable: Option<Counter>,
+}
+
+impl Default for CincinnatiGraphFetchPlugin {
+    fn default() -> Self {
+        Self {
+            upstream: String::from(""),
+            prometheus_registry: None,
+            http_upstream_reqs: None,
+            http_upstream_unreachable: None,
+        }
+    }
+}
+
+impl Debug for CincinnatiGraphFetchPlugin {
+    fn fmt(self: &Self, f: &mut Formatter) -> Result<(), core::fmt::Error> {
+        write!(
+            f,
+            "CincinnatiGraphFetchPlugin {{ upstream: {:?}, prometheus_registry: {} }}",
+            self.upstream,
+            self.prometheus_registry.is_some(),
+        )
+    }
+}
+
+impl PluginSettings for CincinnatiGraphFetchSettings {
+    fn build_plugin(&self) -> Fallible<PluginReference> {
+        let cfg = self.clone();
+        let plugin = CincinnatiGraphFetchPlugin::try_new(cfg.upstream, None)?;
+        Ok(new_pluginref!(InternalPluginWrapper(plugin)))
+    }
+}
+
+impl CincinnatiGraphFetchPlugin {
+    /// Plugin name, for configuration.
+    pub(crate) const PLUGIN_NAME: &'static str = "cincinnati-graph-fetch";
+
+    /// Validate plugin configuration and fill in defaults.
+    pub fn deserialize_config(cfg: toml::Value) -> Fallible<Box<PluginSettings>> {
+        let settings: CincinnatiGraphFetchSettings = cfg.try_into()?;
+
+        ensure!(!settings.upstream.is_empty(), "empty upstream");
+
+        Ok(Box::new(settings))
+    }
+
+    pub fn try_new(
+        upstream: String,
+        prometheus_registry: Option<prometheus::Registry>,
```

Comment:
I think this does not need to be wrapped in an `Option` (or, at least, I didn't spot why you'd need it).

Reply:
I wanted to make the metrics optional, and so the registry isn't always needed.

Reply:
From out-of-band discussion: this is to make testing boilerplate less heavy.